crates/server/src/store/forwarder/authority.rs

```diff
@@ -51,9 +54,41 @@ impl<P: ConnectionProvider> ForwardAuthority<P> {
 
     /// Read the Authority for the origin from the specified configuration
     pub fn try_from_runtime(
+        origin: Name,
+        zone_type: ZoneType,
+        config: &ForwardConfig,
+        runtime: P,
+    ) -> Result<Self, String> {
+        cfg_if! {
+            if #[cfg(feature = "__dnssec")] {
+                Self::with_trust_anchor(origin,zone_type,config,Arc::new(TrustAnchor::default()),runtime)
+            } else {
+                let resolver_constructor = |config, options, runtime| Resolver::new(config, options, runtime);
+                Self::new_inner(origin, zone_type, config, resolver_constructor, runtime)
+            }
+        }
+    }
+
+    /// Construct the Authority from the specified configuration and a set of DNSSEC trust anchors.
+    #[cfg(feature = "__dnssec")]
+    pub fn with_trust_anchor(
+        origin: Name,
+        zone_type: ZoneType,
+        config: &ForwardConfig,
+        trust_anchor: Arc<TrustAnchor>,
+        runtime: P,
+    ) -> Result<Self, String> {
+        let resolver_constructor = |config, options, runtime: P| {
+            Resolver::with_trust_anchor(config, options, trust_anchor, runtime)
+        };
+        Self::new_inner(origin, zone_type, config, resolver_constructor, runtime)
+    }
+
+    fn new_inner(
         origin: Name,
         _zone_type: ZoneType,
         config: &ForwardConfig,
+        resolver_constructor: impl FnOnce(ResolverConfig, ResolverOpts, P) -> Resolver<P>,
```

Comment:
I'm not a fan of this callback stuff. Can we extract a function that builds the `ResolverConfig` from `try_from_runtime()`, or otherwise rearrange the constituent pieces so that there's less indirection here?

Reply:
The biggest problem is that `TrustAnchor` does not always exist, so the closures let me smuggle it in via capturing without writing it in the inner function's signature. I'm going to try introducing a builder API for these instead.